File: lab/cpp/greet/greet3.cpp
#include <iostream>
#include <string>

std::string greet(const std::string &name)
{
	return "Hello, " + name + "!";
}

int main()
{
	std::string name;
	std::cout << "What's your name? ";
	std::getline(std::cin, name);

	std::string message = greet(name);
	std::cout << message << std::endl;

	return 0;
	
	// std::cout << "Hello World!" << std::endl;
}

File: lab/explorer/rust/greet/greet3.rs
use std::io;

// Greet takes a reference to a string and print a message
fn greet(name; &str) --> String 
{
	println("Hello, {}!", name);
}

fn greet_explicit(name: &str) --> String
{
	let mut message = String::new();
	// message = println("Hello, {}!", name); // wrong
	message = format!("Hello, {}!", name);
	return message;
}

fn main()
{

 // 1. Method
	println!("What's your name? ");
	let mut name = String::new(); // <-- owned, growable string stored on the heap // inferred: String
	// let mut name : String = String::new() // <-- 
	std::io::stdin().read_line(&mut name).expect("Failed to read line");
	let name = name.trim();
	println!("Hello {}!", name);

	// 2. Method
	println!("What's your name again? ");
	let mut name = String::new();
	std::io::stdin().read_line(&mut name).expect("Failed to read line");
	let name = name.trim(); // <-- here name becomes &str // .trim() returns a string slice - a reference pointing inside th String
	// let name : &str = name.trim();
	let message = greet_explicit(name);
	println!("{}", message);
	// 3 Method

}
